File: android/app/src/main/java/com/webforge/browser/Personas.kt
package com.webforge.browser

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

data class Persona(val id: String, var name: String, var rules: MutableList<String>, val builtin: Boolean = false)

/**
 * Personas on mobile (#88) — the port of the Windows model (#25). Tabs-only
 * isolation: each Persona owns a tab list, cookies stay shared. URLs matching
 * no rule land in the built-in "Unassigned" Persona so the real ones stay clean.
 *
 * Definitions sync through the same service as bookmarks, so both devices agree.
 */
object Personas {
    const val UNASSIGNED = "unassigned"
    private const val SYNC_URL = "http://100.69.184.113:8013/store/personas"

    private var list: MutableList<Persona>? = null
    private var active: String = UNASSIGNED
    private var updatedAt: Long = 0

    private fun file(c: Context) = File(c.filesDir, "personas.json")

    private fun defaults() = mutableListOf(
        Persona(UNASSIGNED, "Unassigned", mutableListOf(), builtin = true),
        Persona(java.util.UUID.randomUUID().toString(), "Personal", mutableListOf()),
        Persona(java.util.UUID.randomUUID().toString(), "Work", mutableListOf()),
    )

    fun all(c: Context): MutableList<Persona> {
        list?.let { return it }
        val parsed = try {
            val root = JSONObject(file(c).readText())
            updatedAt = root.optLong("updatedAt", 0)
            active = root.optString("active", UNASSIGNED)
            parse(root.optJSONArray("personas") ?: JSONArray())
        } catch (e: Exception) {
            defaults()
        }
        val fixed = if (parsed.isEmpty()) defaults() else parsed
        // Unassigned must always exist.
        if (fixed.none { it.id == UNASSIGNED }) {
            fixed.add(0, Persona(UNASSIGNED, "Unassigned", mutableListOf(), builtin = true))
        }
        list = fixed
        return fixed
    }

    private fun parse(arr: JSONArray): MutableList<Persona> {
        val out = mutableListOf<Persona>()
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val id = o.optString("id")
            if (id.isEmpty()) continue
            val rules = mutableListOf<String>()
            val r = o.optJSONArray("rules")
            if (r != null) for (j in 0 until r.length()) r.optString(j)?.takeIf { it.isNotBlank() }?.let { rules.add(it) }
            out.add(Persona(id, o.optString("name", "Persona"), rules, o.optBoolean("builtin", id == UNASSIGNED)))
        }
        return out
    }

    private fun save(c: Context, push: Boolean = true) {
        val arr = JSONArray()
        for (p in all(c)) {
            arr.put(
                JSONObject().put("id", p.id).put("name", p.name)
                    .put("builtin", p.builtin).put("rules", JSONArray(p.rules))
            )
        }
        try {
            file(c).writeText(
                JSONObject().put("personas", arr).put("active", active)
                    .put("updatedAt", updatedAt).toString()
            )
        } catch (e: Exception) {
        }
        if (push) pushToServer(c)
    }

    /** #87: ordering matches Windows — Unassigned last, it's a fallback. */
    fun ordered(c: Context): List<Persona> {
        val l = all(c)
        return l.filter { it.id != UNASSIGNED } + l.filter { it.id == UNASSIGNED }
    }

    fun get(c: Context, id: String): Persona? = all(c).find { it.id == id }

    fun activeId(c: Context): String {
        all(c)
        return if (get(c, active) != null) active else UNASSIGNED
    }

    fun setActive(c: Context, id: String) {
        if (get(c, id) == null) return
        active = id
        save(c, push = false) // which Persona is showing is a per-device thing
    }

    fun add(c: Context, name: String): Persona? {
        val clean = name.trim()
        if (clean.isEmpty()) return null
        val p = Persona(java.util.UUID.randomUUID().toString(), clean, mutableListOf())
        all(c).add(p)
        updatedAt = System.currentTimeMillis()
        save(c)
        return p
    }

    fun remove(c: Context, id: String): Boolean {
        if (id == UNASSIGNED) return false
        val ok = all(c).removeAll { it.id == id }
        if (ok) {
            if (active == id) active = UNASSIGNED
            updatedAt = System.currentTimeMillis()
            save(c)
        }
        return ok
    }

    fun update(c: Context, id: String, name: String?, rules: List<String>?) {
        val p = get(c, id) ?: return
        if (id != UNASSIGNED && !name.isNullOrBlank()) p.name = name.trim()
        if (rules != null) {
            p.rules = rules.map { it.trim() }.filter { it.isNotEmpty() }.toMutableList()
        }
        updatedAt = System.currentTimeMillis()
        save(c)
    }

    // --- matching: identical semantics to personas.js (prefix / glob / regex) ---
    private val reCache = HashMap<String, Regex?>()

    private fun compile(pattern: String): Regex? = reCache.getOrPut(pattern) {
        val raw = pattern.trim()
        try {
            val m = Regex("""^/(.*)/([a-z]*)$""").find(raw)
            when {
                m != null -> Regex(m.groupValues[1], RegexOption.IGNORE_CASE)
                raw.contains('*') -> {
                    val body = Regex.escape(raw.trimEnd('*')).let { it } // escaped below instead
                    val escaped = raw.trimEnd('*')
                        .replace(Regex("""([.+?^$\{\}()|\[\]\\])"""), "\\\\$1")
                        .replace("*", ".*")
                    Regex("^$escaped", RegexOption.IGNORE_CASE)
                }
                else -> null
            }
        } catch (e: Exception) {
            null // a bad pattern matches nothing rather than breaking routing
        }
    }

    fun matches(url: String, pattern: String): Boolean {
        if (url.isEmpty() || pattern.isBlank()) return false
        compile(pattern)?.let { return it.containsMatchIn(url) && it.find(url)?.range?.first == 0 }
        val p = pattern.trim().trimEnd('*').lowercase()
        return p.isNotEmpty() && url.lowercase().startsWith(p)
    }

    /** Which Persona claims this URL? Unassigned when nothing does. */
    fun forUrl(c: Context, url: String): String {
        for (p in all(c)) {
            if (p.id == UNASSIGNED) continue
            if (p.rules.any { matches(url, it) }) return p.id
        }
        return UNASSIGNED
    }

    // --- sync: definitions are shared, the active selection is not ---------
    fun sync(c: Context, done: (Boolean) -> Unit = {}) {
        Thread {
            var ok = false
            try {
                val conn = URL(SYNC_URL).openConnection() as HttpURLConnection
                conn.connectTimeout = 5000
                conn.readTimeout = 5000
                val body = conn.inputStream.bufferedReader().readText()
                conn.disconnect()
                val root = JSONObject(body)
                val remoteAt = root.optLong("updatedAt", 0)
                val data = root.optJSONArray("data")
                all(c)
                if (data != null && data.length() > 0 && remoteAt > updatedAt) {
                    list = parse(data)
                    updatedAt = remoteAt
                    save(c, push = false)
                    ok = true
                }
            } catch (e: Exception) {
            }
            done(ok)
        }.start()
    }

    private fun pushToServer(c: Context) {
        val arr = JSONArray()
        for (p in all(c)) {
            arr.put(
                JSONObject().put("id", p.id).put("name", p.name)
                    .put("builtin", p.builtin).put("rules", JSONArray(p.rules))
            )
        }
        val stamp = updatedAt
        Thread {
            try {
                val conn = URL(SYNC_URL).openConnection() as HttpURLConnection
                conn.requestMethod = "PUT"
                conn.connectTimeout = 5000
                conn.readTimeout = 5000
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "application/json")
                conn.outputStream.use {
                    it.write(JSONObject().put("data", arr).put("updatedAt", stamp).toString().toByteArray())
                }
                conn.inputStream.use { it.readBytes() }
                conn.disconnect()
            } catch (e: Exception) {
            }
        }.start()
    }
}

File: windows/personas.js
// Personas (#25): named workspaces that tabs are routed into by URL rules.
// Tabs-only isolation — cookies/logins stay shared (user decision 2026-07-29).
// Unmatched URLs land in the built-in "Unassigned" persona so the real ones
// stay clean.
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { app } = require('electron');

const UNASSIGNED = 'unassigned';

let cached = null;
const file = () => path.join(app.getPath('userData'), 'personas.json');

function defaults() {
  return {
    personas: [
      { id: UNASSIGNED, name: 'Unassigned', builtin: true, rules: [] },
      { id: crypto.randomUUID(), name: 'Personal', rules: [] },
      { id: crypto.randomUUID(), name: 'Work', rules: [] },
    ],
    active: UNASSIGNED,
    updatedAt: Date.now(),
  };
}

function load() {
  if (!cached) {
    try {
      cached = JSON.parse(fs.readFileSync(file(), 'utf8'));
    } catch {
      cached = defaults();
      save();
    }
    if (!Array.isArray(cached.personas) || !cached.personas.length) cached = defaults();
    // Unassigned must always exist and must always be first.
    if (!cached.personas.some((p) => p.id === UNASSIGNED)) {
      cached.personas.unshift({ id: UNASSIGNED, name: 'Unassigned', builtin: true, rules: [] });
    }
  }
  return cached;
}

function save() {
  cached.updatedAt = Date.now();
  try {
    fs.writeFileSync(file(), JSON.stringify(cached));
  } catch {}
}

const all = () => load().personas;
const get = (id) => all().find((p) => p.id === id) || null;

function activeId() {
  const a = load().active;
  return get(a) ? a : UNASSIGNED;
}

function setActive(id) {
  if (!get(id)) return false;
  load().active = id;
  save();
  return true;
}

function add(name) {
  const clean = String(name || '').trim();
  if (!clean) return null;
  const p = { id: crypto.randomUUID(), name: clean, rules: [] };
  load().personas.push(p);
  save();
  return p;
}

function remove(id) {
  if (id === UNASSIGNED) return false; // built-in, always present
  const list = load().personas;
  const i = list.findIndex((p) => p.id === id);
  if (i === -1) return false;
  list.splice(i, 1);
  if (cached.active === id) cached.active = UNASSIGNED;
  save();
  return true;
}

function update(id, fields) {
  const p = get(id);
  if (!p) return false;
  if (typeof fields.name === 'string' && fields.name.trim() && id !== UNASSIGNED) {
    p.name = fields.name.trim();
  }
  if (Array.isArray(fields.rules)) {
    p.rules = fields.rules.map((r) => String(r).trim()).filter(Boolean);
  }
  save();
  return true;
}

// Prefix match, case-insensitive, trailing '*' optional — same semantics as
// the tab-group patterns this generalises (#34).
function matches(url, pattern) {
  if (!url || !pattern) return false;
  const p = String(pattern).trim().replace(/\*+$/, '').toLowerCase();
  return p ? String(url).toLowerCase().startsWith(p) : false;
}

/** Which persona owns this URL? Unassigned when no rule claims it. */
function forUrl(url) {
  for (const p of all()) {
    if (p.id === UNASSIGNED) continue;
    if (p.rules.some((r) => matches(url, r))) return p.id;
  }
  return UNASSIGNED;
}

module.exports = { UNASSIGNED, all, get, activeId, setActive, add, remove, update, forUrl, matches };
